crates/red_knot_python_semantic/src/semantic_index/expression.rs

```diff
@@ -44,6 +44,17 @@ pub(crate) struct Expression<'db> {
     #[return_ref]
     pub(crate) node_ref: AstNodeRef<ast::Expr>,
 
+    /// An assignment statement, if this expression is immediately used as the rhs of that
+    /// assignment.
+    ///
+    /// (Note that this is the _immediately_ containing assignment — if a complex expression is
+    /// assigned to some target, only the outermost expression node has this set. The inner
+    /// expressions are used to build up the assignment result, and are not "immediately assigned"
+    /// to the target, and so have `None` for this field.)
+    #[no_eq]
+    #[tracked]
+    pub(crate) assigned_to: Option<AstNodeRef<ast::StmtAssign>>,
```

Comment:
I don't love how this is so specialized to `StmtAssign`. We're not currently creating a standalone expression for annotated assignments, which means we currently raise a false positive for

```py
T: TypeVar = TypeVar("T")
```

Are there any concerns with adding a standalone expression for annotated assignments too?  Then I could make this `ast::Stmt` (or maybe a separate enum that only allows `StmtAssign` and `StmtAnnAssign`)

Reply:
I would be happy with disallowing `T: TypeVar = TypeVar("T")`! I don't think it really makes sense to add the annotation there: it will be ignored by us anyway. It indicates to me that the user is confused about how TypeVars work, and perhaps doesn't realise that a simple `T = TypeVar("T")` assignment will also be considered a declaration by us. Mypy disallows an annotation in this situation: https://mypy-play.net/?mypy=latest&python=3.12&gist=4dad759b00a84071f89fbe9b5e041dcb

I'm also okay with allowing it, though. No strong opinion, as long as we don't emit a _confusing_ error message if somebody does this 😄

Reply:
It would be preferable not to add a standalone expression for every annotated assignment; the more standalone expressions, the more memos, and the slower incremental re-validation gets. If we really need to, we probably can, but I don't think `T: TypeVar = TypeVar("T")` is a sufficiently motivating use case. As Alex says, I think it would be fine to just error on that; `T = TypeVar("T")` is effectively a syntactic special form.

Reply:
> As Alex says, I think it would be fine to just error on that;
> ...
> No strong opinion, as long as we don't emit a _confusing_ error message if somebody does this 😄


The main issue with this is that _customizing_ the error message will require the same extra plumbing as allowing it, since we'd need to detect that the `TypeVar` call occurs in an annotated assignment.  Otherwise we'll have the default error message ``A legacy `typing.TypeVar` must be immediately assigned to a variable``.

Reply:
> Otherwise we'll have the default error message `` A legacy `typing.TypeVar` must be immediately assigned to a variable ``.

Which, to be clear, I'm okay with, certainly for now